File: src/main/java/com/arclights/musiclights/gui/MainWindow.java
package com.arclights.musiclights.gui;

import com.arclights.musiclights.core.LightRig;
import com.arclights.musiclights.gui.light.LightControl;
import com.arclights.musiclights.gui.masterControl.MasterControl;
import javafx.animation.KeyFrame;
import javafx.animation.Timeline;
import javafx.application.Application;
import javafx.scene.Scene;
import javafx.scene.layout.GridPane;
import javafx.scene.paint.Color;
import javafx.stage.Stage;
import javafx.util.Duration;

import java.util.stream.IntStream;

public class MainWindow extends Application {
    private LightRig lightRig;

    private Timeline timeline;

    private final Color[] colors = new Color[]{Color.RED, Color.GREEN, Color.YELLOW,
            Color.BLUE, Color.ORANGE, Color.ORANGE, Color.BLUE, Color.YELLOW,
            Color.GREEN, Color.RED};

    public MainWindow() {
        this.lightRig = new LightRig(10);
    }

    @Override
    public void start(Stage primaryStage) {
        primaryStage.setTitle("MusicLights");
        GridPane root = new GridPane();
        root.setHgap(1);

        Scene scene = new Scene(root);


        // For Arduino
//        Serial s = new Serial(monitor);
//        s.initialize();
//        s.start();

        // The controls
        IntStream.range(0, lightRig.getNbrOfLights()).forEach(i -> {
            LightControl control = new LightControl(i, lightRig, colors[i]);
            root.add(control, i, 0);
        });

        // Master Control
        MasterControl mc = new MasterControl(lightRig, primaryStage);
        root.add(mc, lightRig.getNbrOfLights(), 0);

        root.autosize();

        int drawFrequencyPerSecond = 24;
        timeline = new Timeline(
                new KeyFrame(
                        Duration.seconds(0),
                        event -> lightRig.update()
                ),
                new KeyFrame(Duration.millis(1000 / drawFrequencyPerSecond))
        );
        timeline.setCycleCount(Timeline.INDEFINITE);
        timeline.play();

        primaryStage.setScene(scene);
        primaryStage.sizeToScene();
        primaryStage.show();
    }

    @Override
    public void stop() throws Exception {
        super.stop();
        timeline.stop();
        lightRig.stop();
    }

    public static void main(String[] args) {
        MainWindow.launch(args);
    }

}

File: src/main/kotlin/com/arclights/musiclights/gui/MainWindow.kt
package com.arclights.musiclights.gui

import com.arclights.musiclights.core.LightRig
import com.arclights.musiclights.gui.light.LightControl
import com.arclights.musiclights.gui.masterControl.MasterControl
import javafx.application.Application
import javafx.scene.Scene
import javafx.scene.layout.GridPane
import javafx.scene.paint.Color
import javafx.stage.Stage

class MainWindow : Application() {
    private val colors = arrayOf(Color.RED, Color.GREEN, Color.YELLOW, Color.BLUE, Color.ORANGE)

    private val drawFrequencyPerSecond = 24

    private val lightRig = LightRig(10, 30)

    override fun start(primaryStage: Stage) {
        primaryStage.title = "MusicLights"
        val root = GridPane()
        root.hgap = 1.0

        val scene = Scene(root)

        // The controls
        lightRig.lights.forEachIndexed { index, light ->
            root.add(LightControl(light, colors[index % colors.size]), index, 0)
        }

        // Master Control
        val masterControl = MasterControl(lightRig.config, primaryStage)
        root.add(masterControl, lightRig.nbrOfLights, 0)

        primaryStage.scene = scene
        primaryStage.sizeToScene()
        primaryStage.show()
    }

    override fun stop() {
        super.stop()
        lightRig.stop()
    }

}

fun main(args: Array<String>) {
    Application.launch(MainWindow::class.java, *args)
}